sdk/rust/oak/src/lib.rs

```diff
@@ -154,23 +154,25 @@ pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handl
             ) as i32
         });
         match status {
-            Some(OakStatus::OK) => {
-                unsafe {
-                    buf.set_len(actual_size as usize);
-                    // Handles are written in little-endian order, which matches Wasm spec.
-                    handles.set_len(actual_handle_count as usize);
-                };
-                return OakStatus::OK;
-            }
-            Some(OakStatus::ERR_BUFFER_TOO_SMALL) => {
-                if *resized {
-                    return OakStatus::ERR_BUFFER_TOO_SMALL;
+            Some(s) => match s {
+                OakStatus::OK | OakStatus::ERR_CHANNEL_EMPTY => {
+                    unsafe {
+                        buf.set_len(actual_size as usize);
```

Comment:
Should this use [`resize`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.resize) instead, and avoid the `unsafe`?

Reply:
`resize` didn't seem quite right &ndash; it would need an new value parameter that never gets used, and `set_len` is documented as being  "for situations in which the vector is serving as a buffer for other code, particularly over FFI" which seemed appropriate here.  `resize` also iterates over truncated elements attempting to drop them, which isn't needed.

Reply:
I am still a bit uneasy about this unsafe block, I was looking at it again as part of another refactoring, and I always need to stop and think why it's there and whether it's safe. Would using [`truncate`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate) work better here perhaps? AFAICT this is always going to set the size to be smaller than original value (note that `set_len` would also silently let you increase the size beyond the capacity of the original buffer, if you accidentally passed a higher value). I would not worry about the "dropping" aspect, the compiler will probably optimise those out anyways for such trivial types (though I may be wrong about it).

Reply:
`truncate` changes the length not the capacity; the input vectors here might have (say) length=0 capacity=256 so `truncate` would leave them unchanged.

[Oh, and I guess the same applies to `resize` (which explains why it didn't work when I tried it): using it after the FFI call would overwrite the real data with copies of `resize`'s second argument.]

However, we should definitely improve the commenting around this code to prevent future confusion.

Reply:
#479 for tests and comment improvements.